lld/MachO/ObjC.cpp

```diff
@@ -518,12 +519,32 @@ void ObjcCategoryMerger::collectSectionWriteInfoFromIsec(
 Symbol *
 ObjcCategoryMerger::tryGetSymbolAtIsecOffset(const ConcatInputSection *isec,
                                              uint32_t offset) {
+  if (!isec)
+    return nullptr;
   const Reloc *reloc = isec->getRelocAt(offset);
 
   if (!reloc)
     return nullptr;
 
-  return reloc->referent.get<Symbol *>();
+  Symbol *sym = reloc->referent.get<Symbol *>();
+
+  if (reloc->addend) {
+    assert(isa<Defined>(sym) && "Expected defined for non-zero addend");
+    Defined *definedSym = cast<Defined>(sym);
+    sym = tryFindDefinedOnIsec(definedSym->isec(),
+                               definedSym->value + reloc->addend);
+  }
+
+  return sym;
+}
+
+Defined *ObjcCategoryMerger::tryFindDefinedOnIsec(const InputSection *isec,
+                                                  uint32_t offset) {
+  for (Defined *sym : isec->symbols)
+    if ((sym->value <= offset) && (sym->value + sym->size > offset))
```

Comment:
I wonder why we have an offset in the middle of symbol.

Reply:
I did not fully debug it but it does happen in very rare cases.

Reply:
I am not sure I understand. I can't think of a narrower case that would still actually work.

Reply:
Basically, I'd like to understand why we have an offset in the middle of symbol. If it's unexpected, we should fix the root-cause. If expected, we should document it.

Reply:
OK, I will look into the cases where this happens and come back to this change later.

Reply:
In the test, the addend comes from:

```
__CATEGORY__TtC11SimpleClass11SimpleClass_$_SimpleClass:
	[...]
	.quad	_$s11SimpleClassAACMf+24    // <---- Addend = 24
```

Reply:
The function here seems similar to [this](https://github.com/llvm/llvm-project/blob/f05fa6e0cfdc61f29bac94b157a56e048d10efd1/lld/MachO/InputFiles.cpp#L487). Perhaps, implementation could be shared or refactored?

Reply:
I think they might be different enough to remain separate because: 
- In InputFiles it requires the offset to be to the start of the symbol: 
```
// The offset should point at the exact address of a symbol (with no addend.)
```
Here the offset can be to the inside of the symbol - see [here](https://github.com/llvm/llvm-project/pull/91238/files#diff-71cf08c0f7962a48ce0fef8466a82e328c49072bd05724aae1322bb12313839bR245) 
```
.quad	__DATA__TtC11SimpleClass11SimpleClass+2
```
- In InputFiles it relies on the symbols being ordered - which is not necessarily the case inside optimization passes. So here we have to do a linear scan. I don't think in practice this would be a restriction as the stage of this pass I don't think we run anything that reorders the symbols for these sections - but it's not a hard requirement to have them sorted as far as I know.

Reply:
I know the comment says that but the implementation seems to handle it for the case when it doesn't. Okay with the linear scan, but it would be a nice to make it faster if we know it's already sorted (probably worth checking). If it is sorted, it's worth adding an assert in that case as well to make sure we do do not invalidate correctness behavior.